src/NuGet.Clients/PackageManagement.VisualStudio/Utility/EnvDTEProjectAdapter.cs

```diff
@@ -0,0 +1,305 @@
+            return GetLegacyCSProjProjectReferences(desiredMetadata);
+        }
+
+        public async Task<IEnumerable<LegacyCSProjPackageReference>> GetLegacyCSProjPackageReferencesAsync(Array desiredMetadata)
+        {
+            if (!IsLegacyCSProjPackageReferenceProject)
+            {
+                throw new InvalidOperationException("Package reference call made on a non-legacy CSProj project");
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            return GetLegacyCSProjPackageReferences(desiredMetadata);
+        }
+
+        public async Task AddOrUpdateLegacyCSProjPackageAsync(string packageName, string packageVersion, string[] metadataElements, string[] metadataValues)
+        {
+            if (!IsLegacyCSProjPackageReferenceProject || AsVSProject4 == null)
+            {
+                throw new InvalidOperationException("Cannot add packages to a project which is not a legacy CSProj package reference project");
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            AsVSProject4.PackageReferences.AddOrUpdate(packageName, packageVersion, metadataElements, metadataValues);
+        }
+
+        public async Task RemoveLegacyCSProjPackageAsync(string packageName)
+        {
+            if (!IsLegacyCSProjPackageReferenceProject || AsVSProject4 == null)
+            {
+                throw new InvalidOperationException("Cannot remove packages from a project which is not a legacy CSProj package reference project");
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            AsVSProject4.PackageReferences.Remove(packageName);
+        }
+
+        private IEnumerable<LegacyCSProjProjectReference> GetLegacyCSProjProjectReferences(Array desiredMetadata)
+        {
+            foreach (Reference reference in AsVSProject.References)
+            {
+                if (reference.SourceProject != null)
+                {
+                    yield return new LegacyCSProjProjectReference()
+                    {
+                        UniqueName = reference.SourceProject.FullName
+                        // When metadata API is available, each project's metadata can be inserted into this instance
+                    };
+                }
+            }
+        }
+
+        private IEnumerable<LegacyCSProjPackageReference> GetLegacyCSProjPackageReferences(Array desiredMetadata)
+        {
+            var installedPackages = AsVSProject4.PackageReferences.InstalledPackages;
+            var packageReferences = new List<LegacyCSProjPackageReference>();
+
+            foreach (var installedPackage in installedPackages)
+            {
+                var installedPackageName = installedPackage as string;
+                if (!string.IsNullOrEmpty(installedPackageName))
+                {
+                    string version;
+                    Array metadataElements;
```

Comment:
See earlier comment about intent of adapter. The more transformation that goes downstream to the project, the more testable it is.